### PR TITLE
[perf] Vectorize text-only extend mrope position computation

### PR DESCRIPTION
**Pull request by lcz1998**
## Motivation

`ForwardBatch._compute_mrope_positions` builds the text-only extend branch with a Python list comprehension over every extend token:

```python
mrope_positions = torch.tensor(
    [[pos for pos in range(extend_prefix_len, extend_prefix_len + extend_seq_len)]] * 3
)
```

This sits on the scheduler hot path for every extend/chunked-prefill step of mrope models (Qwen2/2.5/3-VL, Qwen3.5) and costs ~1.4 ms per 4096-token chunk (~2.9 ms at 8192) of pure Python looping per step. We hit it while profiling mixed-chunk serving of Qwen3.5-397B on 8xH800: py-spy showed `_compute_mrope_positions` at ~1.1% of scheduler wall time under a 4k-chunk + 63-decode mixed load.

## Modifications

Replace the per-token list comprehension with a vectorized equivalent:

```python
mrope_positions = (
    torch.arange(extend_prefix_len, extend_prefix_len + extend_seq_len, dtype=torch.int64)
    .unsqueeze(0)
    .expand(3, -1)
)
```

Values, shape `[3, extend_seq_len]`, and int64 dtype are identical; the downstream `torch.cat(...).to(...)` consumes the expanded view unchanged (`cat` materializes it, so no stride-0 view escapes the function). Added a CPU unit test pinning the contract (text-only extend enumeration, mixed text+mm batch concatenation, decode path untouched).

## Accuracy Tests

**Unit test**: `python -m pytest test/registered/unit/model_executor/test_mrope_textonly_extend.py -v` — 3 passed.

**Differential test against the original implementation**: the old method is extracted verbatim from `origin/main` via AST (not a re-typed copy) and both versions are executed end-to-end through `_compute_mrope_positions` — including the downstream `torch.cat(...).to(...)` — on identical inputs. **All 422 cases are bit-exact** (values, shape, dtype=int64):

- exhaustive small grid: prefix 0–8 × extend-len 1–8 (72 cases)
- random large shapes: prefix up to 200k, extend up to 8192, batch size 1–5 (200 cases)
- mixed text-only + multimodal requests in one extend batch (100 cases)
- decode batches, untouched-path sanity (50 cases)

Reproduction script attached in the PR comment below.

**Edge case `extend_seq_len == 0`** (the only input where the two implementations differ):

- Structurally unreachable: `Req._compute_max_prefix_len` caps prefix matching at `input_len - 1` ("the matched length is at most 1 less than the input length to enable logprob computation"), so every request entering an extend batch carries ≥ 1 token.
- If constructed artificially anyway: a single-request batch yields the identical `(3, 0)` int64 result after the downstream cat+to. In a mixed batch, the **old** code's empty tensor is float32 (dtype inferred from the empty Python list) and silently promotes the whole concatenated batch to float32 before the final `.to(torch.int64)`; values survive today only because positions stay < 2^24. The new code stays int64 throughout and removes that hazard.

## Speed Tests and Profiling

Micro-benchmark (CPU, best of 5 runs x 100 iters/call):

| extend_seq_len | old (us/call) | new (us/call) | speedup |
|---:|---:|---:|---:|
| 1024 | 363 | 6.4 | 57x |
| 4096 | 1429 | 7.5 | 190x |
| 8192 | 2855 | 9.1 | 314x |

For a 4096-token chunked prefill step this removes ~1.4 ms of scheduler CPU per step per text-only request being extended.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #29091221340](https://github.com/sgl-project/sglang/actions/runs/29091221340)
Latest PR Test (Extra): :x: [Run #29091221121](https://github.com/sgl-project/sglang/actions/runs/29091221121)

### CODE REVIEW

**Comment by lcz1998** [comment]
Differential-test reproduction script for the accuracy claims in the description (extracts the original `_compute_mrope_positions` verbatim from `origin/main` via AST and compares both implementations end-to-end):

<details>
<summary>difftest_mrope.py (run from repo root: <code>python difftest_mrope.py</code>)</summary>

```python
#!/usr/bin/env python3
"""Differential test: original main-branch `_compute_mrope_positions` vs the
vectorized one, executed end-to-end (including the downstream cat + .to).

The "old" implementation is extracted verbatim from `git show origin/main:...`
via AST, so we compare against the real original code, not a re-typed copy.
"""
import ast
import random
import subprocess
import textwrap
from types import SimpleNamespace

import torch

from sglang.srt.model_executor.forward_batch_info import ForwardBatch, ForwardMode
from sglang.srt.model_executor import forward_batch_info as fbi_module
from sglang.srt.server_args import ServerArgs, set_global_server_args_for_scheduler

set_global_server_args_for_scheduler(ServerArgs(model_path="dummy"))
MODEL_RUNNER = SimpleNamespace(device="cpu")

# -- extract the original method source from git --
blob = subprocess.check_output(
    ["git", "show", "origin/main:python/sglang/srt/model_executor/forward_batch_info.py"],
    text=True,
)
tree = ast.parse(blob)
old_fn_src = None
for node in ast.walk(tree):
    if isinstance(node, ast.ClassDef) and node.name == "ForwardBatch":
        for item in node.body:
            if isinstance(item, ast.FunctionDef) and item.name == "_compute_mrope_positions":
                old_fn_src = ast.get_source_segment(blob, item)
assert old_fn_src is not None, "original method not found"

ns: dict = {
    "torch": torch,
    "get_global_server_args": fbi_module.get_global_server_args,
    "ModelRunner": object,
    "ScheduleBatch": object,
}
exec(textwrap.dedent(old_fn_src), ns)
old_impl = ns["_compute_mrope_positions"]
new_impl = ForwardBatch._compute_mrope_positions
print(f"old impl extracted from origin/main: {len(old_fn_src)} chars")


def make_mm_input(prompt_len: int, delta: int) -> SimpleNamespace:
    return SimpleNamespace(
        mrope_positions=torch.arange(prompt_len, dtype=torch.int64).unsqueeze(0).repeat(3, 1)
        + delta,
        mrope_position_delta=torch.tensor([[delta]], dtype=torch.int64),
        mrope_position_delta_repeated_cache=None,
    )


def run(impl, mode: ForwardMode, reqs: list) -> torch.Tensor:
    """reqs: list of dict(prefix, extend, mm: None | (prompt_len, delta))"""
    fb = ForwardBatch.__new__(ForwardBatch)
    fb.forward_mode = mode
    if mode == ForwardMode.DECODE:
        fb.seq_lens_cpu = torch.tensor([r["seqlen"] for r in reqs])
        batch = SimpleNamespace(
            multimodal_inputs=[
                make_mm_input(*r["mm"]) if r["mm"] else None for r in reqs
            ],
            extend_lens=None,
            prefix_lens=None,
        )
    else:
        fb.seq_lens_cpu = torch.tensor([r["prefix"] + r["extend"] for r in reqs])
        batch = SimpleNamespace(
            multimodal_inputs=[
                make_mm_input(*r["mm"]) if r["mm"] else None for r in reqs
            ],
            extend_lens=[r["extend"] for r in reqs],
            prefix_lens=[r["prefix"] for r in reqs],
        )
    impl(fb, MODEL_RUNNER, batch)
    return fb.mrope_positions


def compare(tag: str, mode: ForwardMode, reqs: list) -> None:
    a = run(old_impl, mode, reqs)
    b = run(new_impl, mode, reqs)
    assert a.shape == b.shape, f"{tag}: shape {a.shape} vs {b.shape}"
    assert a.dtype == b.dtype == torch.int64, f"{tag}: dtype {a.dtype} vs {b.dtype}"
    assert torch.equal(a, b), f"{tag}: values differ"


n = 0
# 1) exhaustive small grid: text-only extend, bs=1
for prefix in range(0, 9):
    for length in range(1, 9):
        compare(f"grid p={prefix} l={length}", ForwardMode.EXTEND,
                [{"prefix": prefix, "extend": length, "mm": None}])
        n += 1

# 2) random large text-only extend, bs 1..5
rng = random.Random(0)
for i in range(200):
    reqs = [
        {"prefix": rng.randint(0, 200_000), "extend": rng.randint(1, 8192), "mm": None}
        for _ in range(rng.randint(1, 5))
    ]
    compare(f"rand-text {i}", ForwardMode.EXTEND, reqs)
    n += 1

# 3) mixed text-only + multimodal requests in one extend batch
for i in range(100):
    reqs = []
    for _ in range(rng.randint(2, 5)):
        if rng.random() < 0.5:
            reqs.append({"prefix": rng.randint(0, 1000), "extend": rng.randint(1, 512), "mm": None})
        else:
            plen = rng.randint(8, 512)
            # prefill within prompt: prefix+extend <= prompt mrope length
            ext = rng.randint(1, plen)
            reqs.append({"prefix": rng.randint(0, plen - ext), "extend": ext,
                         "mm": (plen, rng.randint(-8, 8))})
    compare(f"rand-mixed {i}", ForwardMode.EXTEND, reqs)
    n += 1

# 4) decode batches (untouched path, sanity)
for i in range(50):
    reqs = [
        {"seqlen": rng.randint(1, 100_000),
         "mm": None if rng.random() < 0.5 else (rng.randint(8, 64), rng.randint(-8, 8))}
        for _ in range(rng.randint(1, 5))
    ]
    compare(f"rand-decode {i}", ForwardMode.DECODE, reqs)
    n += 1

print(f"ALL {n} differential cases bit-exact (values, shape, dtype=int64)")

```

</details>

Output on this branch:

```
old impl extracted from origin/main: 2604 chars
ALL 422 differential cases bit-exact (values, shape, dtype=int64)
```

**Comment by lcz1998** [comment]
Hi @hnyls2002 , since you recently touched _compute_mrope_positions in #30708, would you mind taking a look at this small text-only extend optimization when convenient? Thanks!
